waiter/src/waiter/auth/jwt.clj

```diff
@@ -0,0 +1,332 @@
+    (buddy-keys/jwk->public-key entry)))
+
+(defn attach-public-key
+  "Attaches the EdDSA public key into the provided entries."
+  [entry]
+  (assoc entry ::public-key (retrieve-public-key entry)))
+
+(defn retrieve-jwks-with-retries
+  "Retrieves the JWKS using the specified url.
+   JWKS retrieval tried retry-limit times at intervals on retry-interval-ms ms when there is an error."
+  [http-client url
+   {:keys [retry-interval-ms retry-limit]
+    :or {retry-interval-ms 100
+         retry-limit 2}
+    :as options}]
+  (let [with-retries (utils/retry-strategy
+                       {:delay-multiplier 1.0, :initial-delay-ms retry-interval-ms, :max-retries retry-limit})
+        http-options (dissoc options :retry-interval-ms :retry-limit)]
+    (with-retries
+      (fn retrieve-jwks-task []
+        (if (str/starts-with? url "file://")
+          (-> url slurp json/read-str walk/keywordize-keys)
+          (pc/mapply hu/http-request http-client url http-options))))))
+
+(defn refresh-keys-cache
+  "Update the cache of users with prestashed JWK keys."
+  [http-client http-options url supported-algorithms keys-cache]
+  (metrics/with-timer!
+    (metrics/waiter-timer "core" "jwt" "refresh-keys-cache")
+    (fn [elapsed-nanos]
+      (log/info "JWKS keys retrieval took" elapsed-nanos "ns"))
+    (let [response (retrieve-jwks-with-retries http-client url http-options)]
+      (when-not (map? response)
+        (throw (ex-info "Invalid response from the JWKS endpoint"
+                        {:response response :url url})))
+      (let [all-keys (:keys response)
+            keys (filter #(supported-key? supported-algorithms %) all-keys)]
+        (when (empty? keys)
+          (throw (ex-info "No supported keys found from the JWKS endpoint"
+                          {:response response
+                           :url url})))
+        (log/info "retrieved entries from the JWKS endpoint" response)
+        (reset! keys-cache {:key-id->jwk (->> keys
+                                           (map attach-public-key)
+                                           (pc/map-from-vals :kid))
+                            :last-update-time (t/now)
+                            :summary {:num-filtered-keys (count keys)
+                                      :num-jwks-keys (count all-keys)}})))))
+
+(defn start-jwt-cache-maintainer
+  "Starts a timer task to maintain the keys-cache."
+  [http-client http-options jwks-url update-interval-ms supported-algorithms keys-cache]
+  {:cancel-fn (du/start-timer-task
+                (t/millis update-interval-ms)
+                (fn refresh-keys-cache-task []
+                  (refresh-keys-cache http-client http-options jwks-url supported-algorithms keys-cache)))
+   :query-state-fn (fn query-jwt-cache-state []
+                     @keys-cache)})
+
+(def ^:const bearer-prefix "Bearer ")
+
+;; access the private function for validating claims
+(def ^:const validate-claims #'jwt/validate-claims)
```

Comment:
Can we avoid using a private function so that we can feel safe about upgrading?

Reply:
I'd rather write our own claims validation function than use this private function.

Reply:
We shouldn't implement hacks to depend on private methods of outside projects.  Please implement the function ourselves or find another way to get that method called via methods that are available to us publicly.  Let's not leave this tech debt for ourselves.

Reply:
Implemented our own `validate-claims`